### PR TITLE
Properly sanitize student name, team name and comments related to StudentAttributes in the front end #4806

### LINKED ISSUE #4806: Properly sanitize student name, team name and comments related to StudentAttributes in the front end
I will look through all JSON files that have `students` and append some special HTML characters to  `name`, `lastName`, `comments` and `team`fields to make sure that nowhere else is left out.
Any recommendation is appreciated.

### PR DESCRIPTION
**Pull request by bamboo3250**
Fixes #4806

Still Ongoing...

@kanghj this is going to be a long PR :grin:

### CODE REVIEW

**Comment by kanghj** [comment]
Let's change the issue and pr title to reflect that this pr is also going to fix the parts of the code that didn't handle sanitization correctly

File: src/test/resources/pages/instructorSearchPageSearchComments.html [hunk 1]
```diff
@@ -105,7 +105,7 @@
               <span class="text-muted">
                 To
                 <b>
-                  student1 In Course1 (Team 1.1, searchUI.student1InCourse1@gmail.tmt)
+                  student1 In Course1\u003c/option\u003e\u003c/td\u003e\u003c/div\u003e\u0027\" (Team 1.1\u003c/td\u003e\u003c/div\u003e\u0027\", searchUI.student1InCourse1@gmail.tmt)
```
Comment (bamboo3250) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: @kanghj Just a note for you. This part is sanitized properly but these unicode is due to we save them via `getTeammatesGson()` or just `new Gson()` without using `disableHtmlEscaping()`. Currently, only searching comments in Search Page is affected. I'll fix it in another PR.

**Comment by wkurniawan07** [comment]
Merged #4910, and as expected there're conflicts here :v:

File: src/main/java/teammates/common/util/StringHelper.java [hunk 2]
```diff
@@ -373,7 +373,7 @@ public static String csvToHtmlTable(String str) {
             
             result.append("<tr>");
             for (String td : rowData) {
-                result.append(String.format("<td>%s</td>\n", td));
+                result.append(String.format("<td>%s</td>\n", Sanitizer.sanitizeForHtml(td)));
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: By doing this, #3622 is resolved.

**Comment by bamboo3250** [comment]
@kanghj ready for review.

File: src/test/resources/pages/instructorSearchPageSearchComments.html (hunk 1 above)
Comment (kanghj) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: Does this need to be fixed first before we can proceed with this pr?

File: src/main/java/teammates/ui/controller/InstructorCourseEnrollPageData.java [hunk 3]
```diff
@@ -11,8 +11,8 @@
     
     public InstructorCourseEnrollPageData(AccountAttributes account, String courseId, String enrollStudents) {
         super(account);
-        this.courseId =  sanitizeForHtml(courseId);
-        this.enrollStudents = sanitizeForHtml(enrollStudents);
+        this.courseId =  courseId;
+        this.enrollStudents = enrollStudents;
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorCourseEnrollPageData.java]: Where is this sanitized now? I can't find it

File: src/test/resources/pages/instructorSearchPageSearchComments.html (hunk 1 above)
Comment (bamboo3250) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: We can fix it later. In Search Page, we only have to show search results so this bug only affect frontend. It is not critical.

File: src/main/java/teammates/ui/controller/InstructorCourseEnrollPageData.java (hunk 3 above)
Comment (bamboo3250) [inline src/main/java/teammates/ui/controller/InstructorCourseEnrollPageData.java]: It is not shown here because it has already been fixed in #4849

File: src/main/java/teammates/common/datatransfer/CommentAttributes.java [hunk 4]
```diff
@@ -213,14 +212,6 @@ public void sanitizeForSaving() {
         this.courseId = Sanitizer.sanitizeForHtml(courseId);
         this.giverEmail = Sanitizer.sanitizeForHtml(giverEmail);
 
-        if (recipients != null) {
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: This means we have comments that are html-sanitized in the db too?Do we need a script to unsanitise these too?

Comment (bamboo3250) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: Yes, we need another script for `CommentAttributes` too. To be safe, perhaps I will write a script to check for `recipients` of `CommentAttributes` in database.

Comment (bamboo3250) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: Ah wait, maybe I should revert this part and assume `recipients` has been already sanitized and temporarily remove `escapeXml()` in the frontend.

File: src/main/java/teammates/common/datatransfer/CommentAttributes.java [hunk 5]
```diff
@@ -213,14 +212,6 @@ public void sanitizeForSaving() {
         this.courseId = Sanitizer.sanitizeForHtml(courseId);
         this.giverEmail = Sanitizer.sanitizeForHtml(giverEmail);
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: if the recipients are changed, looks like this should be changed too

File: src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperTwo.html [hunk 6]
```diff
@@ -315,16 +315,16 @@
             <tbody>
               <tr>
                 <td class="middlealign">
-                  Team 1
+                  Team 1</td></div>'"
```
Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperTwo.html]: Just to verify, to see if this is correctly handled, it's based on the indentation level right?

File: src/main/java/teammates/common/datatransfer/CommentAttributes.java (hunk 5 above)
Comment (bamboo3250) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: It is email so I doubt it would get affected by sanitization.

File: src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperTwo.html (hunk 6 above)
Comment (bamboo3250) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperTwo.html]: That's right. The browser unsanitizes those strings when displaying to us. If it's not correctly handled, the structure will be messed up (i.e. team name is broken into 2 parts) and we can see it immediately.

File: src/main/java/teammates/common/datatransfer/CommentAttributes.java (hunk 4 above)
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: Yes, perhaps we will have to do this in another issue to handle sanitization for comments then...

File: src/test/resources/pages/instructorFeedbackResultsPageOpenViewForHelperTwo.html [hunk 7]
```diff
@@ -2625,9 +2625,7 @@
           <tbody>
             <tr>
               <td>
-                <i>
-                  Instructors
-                </i>
+                <i>Instructors</i>
```
Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackResultsPageOpenViewForHelperTwo.html]: It's strange that this changed

Comment (bamboo3250) [inline src/test/resources/pages/instructorFeedbackResultsPageOpenViewForHelperTwo.html]: I also wonder why but cannot find the reason.

**Comment by bamboo3250** [comment]
@kanghj I just recovered changes in CommentAttributes and convert un/sanitized strings to their correct state. It may be confusing though.

File: src/main/java/teammates/ui/controller/StudentCommentsPageData.java [hunk 8]
```diff
@@ -149,13 +150,14 @@ private void setCommentsForStudentsTables(String studentEmail, CourseRoster rost
...
+            String unsanitizedRecipientDetails = StringHelper.recoverFromSanitizedText(recipientDetails);
             InstructorAttributes instructor = roster.getInstructorForEmail(comment.giverEmail);
             String giverDetails = comment.giverEmail;
             if (instructor != null) {
                 giverDetails = instructor.displayedName + " " + instructor.name;
             }
-            
-            Comment commentRow = new Comment(comment, giverDetails, recipientDetails);
+            String unsanitizedGiverDetails = StringHelper.recoverFromSanitizedText(giverDetails);
+            Comment commentRow = new Comment(comment, unsanitizedGiverDetails, unsanitizedRecipientDetails);
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/StudentCommentsPageData.java]: Where does this get sanitized now? It's very hard to look for files on the new github pr page.

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag [hunk 9]
```diff
@@ -202,10 +202,10 @@
...
-    <input type="hidden" name="<%= Const.ParamsNames.FEEDBACK_RESPONSE_ID %>" value="${frc.feedbackResponseId}">
+    <input type="hidden" name="<%= Const.ParamsNames.FEEDBACK_RESPONSE_ID %>" value="${fn:escapeXml(frc.feedbackResponseId)}">
     <input type="hidden" name="<%= Const.ParamsNames.COURSE_ID %>" value="${frc.courseId}">
     <input type="hidden" name="<%= Const.ParamsNames.FEEDBACK_SESSION_NAME %>" value="${frc.feedbackSessionName}">
     <input type="hidden" name="<%= Const.ParamsNames.USER_ID %>" value="${data.account.googleId}">
-    <input type="hidden" name="<%= Const.ParamsNames.RESPONSE_COMMENTS_SHOWCOMMENTSTO %>" value="${frc.showCommentToString}">
-    <input type="hidden" name="<%= Const.ParamsNames.RESPONSE_COMMENTS_SHOWGIVERTO %>" value="${frc.showGiverNameToString}">
+    <input type="hidden" name="<%= Const.ParamsNames.RESPONSE_COMMENTS_SHOWCOMMENTSTO %>" value="${fn:escapeXml(frc.showCommentToString)}">
+    <input type="hidden" name="<%= Const.ParamsNames.RESPONSE_COMMENTS_SHOWGIVERTO %>" value="${fn:escapeXml(frc.showGiverNameToString)}">
```
Comment (kanghj) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: How come this needs to be escaped?

File: src/test/java/teammates/test/cases/common/CommentAttributesTest.java [hunk 10]
```diff
@@ -107,29 +106,6 @@ public void testValidate() {
         assertEquals(expectedErrorMessage.toString(), errorMemssage.toString());
     }
     
-    @Test
-    public void testSanitize() {
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/common/CommentAttributesTest.java]: Should this be reverted too?

**Comment by kanghj** [comment]
Some comments added. There are merge conflicts now too.

File: src/main/java/teammates/common/util/Sanitizer.java [hunk 11]
```diff
@@ -121,6 +123,28 @@ public static String sanitizeForHtml(String str){
     }
     
     /**
+     * Sanitize a list of strings for inserting into HTML.
```
Comment (damithc) [inline src/main/java/teammates/common/util/Sanitizer.java]: Sanitizes

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 12]
```diff
@@ -926,7 +926,7 @@ public void testGetFeedbackSessionResultsForUser() throws Exception {
                 "FSRTest.student6InCourse1@gmail.tmt=student6 In Course1",
                 "FSRTest.instr1@course1.tmt=Instructor1 Course1",
                 "FSRTest.instr2@course1.tmt=Instructor2 Course1",
-                "Anonymous student 283462789@@Anonymous student 283462789.com=Anonymous student 283462789",
+                "Anonymous student 1479876785@@Anonymous student 1479876785.com=Anonymous student 1479876785",
```
Comment (damithc) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: Is this change intended?

**Comment by damithc** [comment]
Some cosmetic changes suggested. Can merge after that.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java (hunk 12 above)
Comment (bamboo3250) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: Yes, it is hashcode of the student's name so if the student name is changed, the hashcode should be changed as well.

**Comment by bamboo3250** [comment]
I'm merging this PR now.

File: src/test/resources/pages/instructorSearchPageSearchComments.html (hunk 1 above)
Comment (bamboo3250) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: @kanghj i mean here

File: src/test/resources/pages/instructorFeedbackResultsPageOpenViewForHelperTwo.html (hunk 7 above)
Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackResultsPageOpenViewForHelperTwo.html]: This looks like a mistake now, @bamboo3250.
